src/topology/homotopy/H_spaces.lean

```diff
@@ -0,0 +1,453 @@
+/-
+Copyright (c) 2022 Filippo A. E. Nuccio Mortarino Majno di Capriglio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio
+-/
+import data.real.basic
+import tactic.field_simp
+import topology.algebra.order.basic
+import topology.compact_open
+import topology.homotopy.basic
+import topology.homotopy.path
+
+universes u v w
+
+classical theory
+
+noncomputable theory
+
+open_locale unit_interval
+
+namespace unit_interval
+
+@[simp]
+lemma univ_eq_ge_zero {θ : ℝ} (h : θ ≤ 0): {s : I | θ ≤ (s : ℝ)} = set.univ :=
+begin
+  rw set.eq_univ_iff_forall,
+  intro,
+  simpa only [set.mem_set_of_eq] using (h.trans $ nonneg _),
+end
+
+@[simp]
+lemma univ_eq_le_one {θ : ℝ} (h : 1 ≤ θ) : {s : I | (s : ℝ) ≤ θ} = set.univ :=
+begin
+  rw set.eq_univ_iff_forall,
+  intro,
+  simpa only [set.mem_set_of_eq] using (le_one _).trans h,
+end
+
+lemma mem_closure_iff {θ : ℝ} {t : I} : t ∈ closure {s : I | (s : ℝ) ≤ θ} ↔ (t : ℝ) ≤ θ :=
+  by {rw [(is_closed_le continuous_induced_dom continuous_const).closure_eq, set.mem_set_of_eq],
+    apply_instance}
+
+lemma not_mem_interior {θ : ℝ} {t : I} : t ∉ interior {s : I | (s : ℝ) ≤ θ} → θ ≤ t :=
+begin
+  -- For `θ = 1` the set `{s : ↥I | (s : ℝ) ≤ θ}` is the whole `I`, whose interior (in the induced
+  -- topology) is `I` again, and `t ∉ interior I = I` is always false, whereas `1 ≤ t` is true for
+  -- `t = 1`. So we split the proof in the cases `1 ≤ θ` and `θ < 1`.
+  by_cases h_θ_1 : 1 ≤ θ,
+  { rw [univ_eq_le_one h_θ_1, interior_univ],
+    tauto },
+  intro h,
+  by_cases h_θ_0 : 0 ≤ θ,
+  { let θI : I := ⟨θ, set.mem_Icc.mpr ⟨h_θ_0, le_of_lt $ lt_of_not_ge h_θ_1⟩⟩,
+    have : {s : ↥I | (s : ℝ) ≤ θ} = set.Iic θI := rfl,
+    have H_ne : (set.Ioi θI).nonempty := ⟨1, _⟩,
+    simpa only [this, interior_Iic' H_ne, ← set.Iio_def, set.mem_set_of_eq, not_lt,
+      ← subtype.coe_le_coe] using h,
+    simpa only [set.mem_Ioi, ← subtype.coe_lt_coe, subtype.coe_mk, set.Icc.coe_one] using
+      lt_of_not_ge h_θ_1, },
+  { exact le_of_lt (lt_of_lt_of_le (lt_of_not_ge h_θ_0) t.2.1) },
+end
+
+lemma mem_frontier {θ : ℝ} {t : I} : t ∈ frontier (λ i : I, (i : ℝ) ≤ θ) → (t : ℝ)
+  = θ := λ ⟨h_left, h_right⟩, by {simp only [eq_of_ge_of_not_gt (not_mem_interior
+    h_right) (not_lt_of_le $ mem_closure_iff.mp h_left)]}
+
+end unit_interval
+
+
+namespace path
+
+open continuous_map path
+
+variables {X : Type u} [topological_space X]
+
+instance {x y : X} : has_coe (path x y) C(I, X) := ⟨λ γ, γ.1⟩
+
+instance {x y : X} : topological_space (path x y) :=
+  topological_space.induced (coe : _ → C(↥I, X)) continuous_map.compact_open
+
+lemma continuous_eval {x y : X} : continuous (λ p : I × path x y, p.2 p.1) :=
+(continuous_eval'.comp $ continuous_induced_dom.prod_map (@continuous_id I _)).comp continuous_swap
+
+end path
+
+namespace H_space
```

Comment:
You entered namespace too early, so that `class H_space` below will get the name `H_space.H_space`.